### ISSUE TITLE
ReLoRA prunes adapters without merging and can zero both factors on ties

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#693]
## Summary

Soup's feature named ReLoRA periodically magnitude-prunes the learned `lora_A` and `lora_B` weights without first merging their update into the base model. This is not the merge-and-reinitialize restart defined by ReLoRA and it abruptly changes the learned function.

There is also a concrete pruning bug: strict thresholding can remove every entry when magnitudes tie, even though `prune_ratio < 1`. If both LoRA factors become zero, the branch receives zero ordinary gradients and cannot restart.

## Current behavior

On current `main` (`ef2e02b51a1b34546823edce55715d28234805cb`), `_prune_and_reset()`:

1. selects parameters whose names contain `lora_A` or `lora_B`;
2. calls `magnitude_prune_tensor()` on those adapter weights;
3. clears optimizer state for those same parameters.

It never merges `B @ A` into the frozen base weight and never reinitializes A/B in the canonical LoRA shape.

The ReLoRA paper's restart instead merges the low-rank update into the base, reinitializes A with Kaiming and B with zeros, prunes/resets optimizer state, and restarts the learning-rate warmup: https://arxiv.org/html/2307.05695#S2

Soup's module docstring says it mirrors the paper/Axolotl, while the public docs describe adapter-weight pruning. That makes this both an algorithm-conformance problem and a behavior that is destructive to the currently learned adapter contribution.

## Reproductions

Using the actual callback kernel from the pinned checkout:

### Learned-function discontinuity

With random non-zero LoRA factors, the base weight is unchanged by the callback and the MSE between the pre-callback and post-callback effective linear maps changes from 0 to `3.739178...`.

A merge-then-reset control preserves the effective map exactly.

This demonstrates function modification, not downstream model-quality impact; no full LLM training was run.

### Tie handling can kill the branch

The mask is:

```python
threshold = torch.kthvalue(flat, k).values
mask = tensor.detach().abs() > threshold
```

All values equal to the threshold are removed. With constant A and B tensors and `prune_ratio=0.9`, the actual implementation yields:

```text
nonzero A: 0
nonzero B: 0
MSE after callback: 16.0
gradient norm A: 0.0
gradient norm B: 0.0
```

The equal-magnitude case is constructed; this report does not claim a measured production frequency. The exact-count violation itself is deterministic.

## Recommended fix

Choose and document one of two coherent contracts:

1. Implement ReLoRA restarts: merge the adapter delta into the base weight, reinitialize A/B, reset or magnitude-prune the optimizer moments (not the learned adapter weights), and re-warm the LR schedule; or
2. Rename the current feature to periodic LoRA weight pruning and remove claims that it mirrors ReLoRA.

For a canonical implementation, quantized bases, FSDP/DeepSpeed wrappers, layer streaming, and checkpoint/resume need explicit support or a loud compatibility rejection.

Independently, replace threshold comparison with exact index selection (for example a deterministic `topk`/stable-order mask) so the requested keep count is honored under ties and at least one entry survives in each non-trivial factor.

## Acceptance criteria

- [ ] A restart preserves the effective linear map across merge/reinitialization to numerical tolerance.
- [ ] ReLoRA restarts accumulate successive low-rank updates in the base model.
- [ ] Optimizer-state pruning/reset targets optimizer moments, not learned adapter weights.
- [ ] Tied magnitudes retain exactly the requested number of entries deterministically.
- [ ] A regression test proves both A and B cannot be zeroed into a zero-gradient branch by `prune_ratio < 1`.
- [ ] Unsupported quantized/distributed/streamed combinations fail clearly rather than partially applying a restart.

## Duplicate check

#67 covers wiring ReLoRA into additional trainers; #308 fixes callback inheritance; #320/PR #455 fixes lazy imports. None covers missing merge/reinitialization or tie-induced total pruning. Searches across open/closed issues and PRs found no matching report or open fix.

**Comment by MakazhanAlpamys** [comment]
**Confirmed — both halves, checked separately, because one could easily have held without the other.**

Real `ReLoRACallback._prune_and_reset` against a real peft LoRA model on `2edc063`:

```
CLAIM A - merge before prune?
  base weight changed by callback : False   (max|delta| = 0.000e+00)
  MSE(effective map before, after): 142.120667
  CONTROL merge_and_unload MSE    : 8.771e-13

CLAIM B - ties, constant-magnitude factors, prune_ratio=0.9
  lora_A  numel=32  nonzero=0   (requested keep = 4)
  lora_B  numel=32  nonzero=0   (requested keep = 4)
  grad-norm lora_A : 0.000000
  grad-norm lora_B : 0.000000

  kernel: ones(10)      prune_ratio=0.9 -> nonzero=0  (expected 1)
  kernel: arange(1..10) prune_ratio=0.9 -> nonzero=1  (expected 1)
```

**Claim A holds.** The base weight is provably untouched — `max|delta| = 0` — while the effective linear map moves by MSE 142.12. The merge-then-reset control moves it by 8.77e-13, i.e. not at all. So the adapter's learned contribution is *discarded*, not restarted. That is the conformance gap you name against the module's own docstring at `relora.py:3-5`, which claims to mirror the paper and Axolotl.

**Claim B holds, and the gradient consequence is real rather than theoretical.** `magnitude_prune_tensor` clamps `k >= numel` to `numel-1`, which rescues the non-tied case — `arange` keeps 1 survivor, correct — but the mask is `abs(t) > threshold`, strictly greater, so a constant tensor is wiped entirely. With both A and B at zero, `B @ A @ x` returns zero gradient to both factors, measured at `0.000000` each. The branch is permanently dead, exactly as you describe, and the exact-count violation is deterministic: 0 kept where 4 were requested.

## Two notes for whoever fixes it

**Claim A is a maintainer decision as much as a patch.** Fixing it means either implementing real restarts — merge `B@A` into the base, then reinitialise A and B — or renaming the feature and correcting the docstring so it stops claiming conformance with the paper. I would rather have the first, but the second is honest and much smaller, and I do not want someone doing half of the first.

**Claim B's production frequency is worth a thought before fixing only the constructed case.** You label the tie scenario constructed, which is fair — but a freshly-initialised `lora_B` is **all zeros** by PEFT convention. That is the natural tie, and it is worth checking whether the first `relora_steps` interval after initialisation hits it.

Severity: anyone setting `training.relora_steps`, SFT on the transformers backend only. Claim A is silent and fires on **every** ordinary run, not just an edge case — which makes it the more consequential of the two despite Claim B being the more dramatic.

Labelled `bug` + `help wanted`, unclaimed. Comment here before opening a PR, and please say which of the two you are taking, since A and B are separable.

**Comment by Srinivasan8888** [comment]
**Taking this — both halves, and I reproduced each independently before claiming rather than working from your table.**

**Overlap check first.** No claim on this issue. `gh pr list --state all --search` for relora/prune returns nothing touching `utils/relora.py`; the open PRs are #746 (eval judge numerics), #742/#743/#736 (distill), #741/#737 (PPO), #740/#739/#734 (MLX), #730/#729/#727/#726/#718, #660, #592 — none in this file. Nothing of mine is near it either.

## Claim B, reproduced

```
  ones(10)            numel= 10  requested_keep= 1  nonzero_after= 0   <-- WIPED
  arange(1..10)       numel= 10  requested_keep= 1  nonzero_after= 1
  full(10, -3.0)      numel= 10  requested_keep= 1  nonzero_after= 0   <-- WIPED
  ones(4,8) [lora]    numel= 32  requested_keep= 4  nonzero_after= 0   <-- WIPED

  ones(10): threshold = kthvalue(flat, 9) = 1.0  ->  |t| > 1.0 is False everywhere
```

and the gradient consequence, measured rather than argued:

```
  after pruning both:  A nonzero=0  B nonzero=0
  grad-norm lora_A  = 0.000000
  grad-norm lora_B  = 0.000000
```

**Your open question has an answer, and it is the bad one.** You asked whether a freshly-initialised `lora_B` is the natural tie. Against a real PEFT model (`sshleifer/tiny-gpt2`, `target_modules=["c_attn"]`):

```
  lora_A.default   numel=8   distinct_values=8
  lora_B.default   numel=24  distinct_values=1   ALL-EQUAL (tie)
```

So the tie is PEFT's own initialisation, not a constructed case. That moves B from "dramatic but rare" to "the shape every run starts in".

## Claim A, reproduced

```
  base weight changed by the callback : False   (max|delta| = 0.000e+00)
  MSE(effective map before, after)    : 2.106e-07     max|logit delta| = 4.871e-03
  CONTROL merge_and_unload MSE        : 5.126e-18
```

Base weight provably untouched, map moves, control does not — eleven orders apart. Smaller absolute numbers than yours because tiny-gpt2 is 2-dimensional and I scaled `lora_B` to 0.1; the ratio is the same story.

**Two corrections to my own measurement, since they would have made me report noise as signal.** My first Claim A run gave MSE 4.75e-04 with a *control* of 4.59e-04 — the control should be ~0, so the measurement was wrong, not the code. Cause: I had not called `.eval()`, so GPT-2 dropout made every forward stochastic and I was measuring dropout. In eval mode the control is 5.126e-18. I mention it because your 8.77e-13 is the number that told me mine was broken.

## What I am taking, and the one thing I want your call on

**Both**, as separable commits in one PR: B first (a strict `>` that must be `>=`, plus the exact-count guarantee, which is a small change with a real test), then A as **real restarts** — merge `B@A` into the base layer, then reinitialise A and B — which is the option you said you would rather have.

The reason I would rather not take A alone or B alone: after a correct restart, PEFT reinitialises `lora_B` to zeros, which is precisely B's all-equal input. So A's fix hands B's bug a fresh instance on the very next fire. I have not yet measured whether that is harmful or merely a no-op there — `B = 0` means `B@A@x = 0` and the gradient to `A` is zero, but the gradient to `B` is not, which is ordinary PEFT behaviour and recovers. I will pin whichever it turns out to be rather than assert it now.

**Where I would like your steer:** for A, `merge_and_unload()` collapses the adapter permanently, so a restart needs a per-layer merge that keeps the LoRA modules attached (`layer.merge()` / `layer.unmerge()` on PEFT's `LoraLayer`, or writing `B@A * scaling` into `base_layer.weight` directly). If you have a preference between using PEFT's own merge API and doing the arithmetic in Soup, say so — the first is less code and tracks upstream, the second does not depend on PEFT keeping `merge()` public. I lean on PEFT's API.

CPU-only, no GPU needed for either half, so I can verify the acceptance criteria on this machine.

**Comment by MakazhanAlpamys** [comment]
@SatvikMishra08 — checking in on the hold I placed here for you on 2026-09-06.

No pressure either way, but I would rather ask than leave it parked silently: are you still interested in this one? It has been seven days with no branch or PR against it.

Being straight about the cost, because it is not zero: I turned @jagadeepmamidi away from this work to keep the hold, and they took #720 and then #367 instead. If I do not hear from you in a couple of days I will release both #692 and #693, and @jagadeepmamidi has first refusal since they stood down for this.

If it has gone cold, just say so — that is a completely fine answer and costs you nothing here. Silence is the only version that costs someone else.

**Comment by MakazhanAlpamys** [comment]
@SatvikMishra08 — I said on 2026-09-13 that I would release this "in a couple of days" if I did not hear back. That was five days ago and I did not follow through, so this is the follow-through rather than a new deadline.

**Are you still on it?** A yes with no code is fine — say the word and it stays yours. Silence for another **three days (through 2026-09-21)** and I will mark it unclaimed and offer it to @jagadeepmamidi, who asked for it and was told you had it first.

I want to be clear that the cost here is not yours to carry alone: somebody was turned away to keep this reserved for you, and then I let it sit. Both halves of that are worth fixing.

If you are stuck on something specific, say what — ReLoRA pruning without merging is genuinely fiddly, and "I got into it and it is harder than it looked" is a perfectly good answer that costs you nothing here.

**Comment by MakazhanAlpamys** [comment]
The three-day window I set on 2026-09-18 ends today, and there has been no reply, so I am doing what
I said I would do rather than extending it again: **#693 is unclaimed as of now.**

@SatvikMishra08 — no criticism intended and none implied. Plans change and an unanswered issue costs
you nothing; the only thing I owe you is that the deadline I published is the deadline I act on. If
you come back to this later and it is still open, it is yours again for the asking.

@jagadeepmamidi — you asked for this one earlier and I turned you down because it was claimed. It is
open now, and it is yours if you still want it. Two things worth knowing before you start:

- The defect is in the ReLoRA pruning path (`utils/relora.py`): adapters are pruned without being
  merged first, so the pruned magnitude is discarded rather than folded into the base, and a tie
  between the two factors can zero both — which silently turns the adapter off for the rest of the
  run instead of resetting it.
- The acceptance criteria on this issue are unchanged. What I will ask for at review is the same
  thing I ask everyone: a test that fails on `main` and passes on your branch, and a mutation of
  your own fix that makes your new test fail. If the merge-before-prune ordering is the fix, the
  mutation is "prune without merging" and it must redden something.

Say the word here and I will assign it to you.

**Comment by jagadeepmamidi** [comment]
Yes @MakazhanAlpamys assign this to me I'll work on it.

**Comment by MakazhanAlpamys** [comment]
Assigned — it is yours, @jagadeepmamidi, and thank you for being patient about the order rather than
about the issue.

One practical note before you start, from #1071: `main` moves under branches here several times a
day, so re-derive rather than assume when you rebase. And the acceptance bar is the one on the issue
plus the standing one — a test that fails on `main`, and a mutation of your own fix that reddens your
new test. If merge-before-prune is the fix, "prune without merging" is the mutation and it has to
break something by name.

The measurement in my 2026-09-05 comment is the starting point: it checks both halves separately
(does the base weight change before pruning, and can a tie zero both factors), because one of them
could hold without the other.

**Comment by jagadeepmamidi** [comment]
Taking the paper-restart contract (not the rename): merge B@A into base, reinit A/B, full optimizer clear, LR re-warmup. Schema loud-rejects quant/stream_layers/VeRA/DoRA/FSDP2 compile; wrapped optimizers without a \.state\ dict fail before merge. Branch pushed to my fork: \ix/issue-693-relora-restart\ — PR incoming. Includes tests that fail on main (effective-map continuity, tie keep-count, merge-skipped mutation) plus Bugbot follow-ups.
